src/libpmemfile-posix/data.c

```diff
@@ -0,0 +1,1332 @@
+
+			/* How many bytes in this hole can be used? */
+			uint64_t hole_count = next->offset - offset;
+
+			/* Are all those bytes needed? */
+			if (hole_count > size)
+				hole_count = size;
+
+			if (hole_count > 0) { /* Is there any hole at all? */
+				block = block_list_insert_after(vinode, block);
+				block->offset = offset;
+				file_allocate_block_data(pfp, block, hole_count,
+				    false);
+				block_cache_insert_block(vinode->blocks, block);
+
+				if (block->size > hole_count)
+					block->size = (uint32_t)hole_count;
+			} else {
+				block = next;
+			}
+		}
+	} while (size > 0);
+}
+
+static struct pmemfile_block *
+find_following_block(PMEMfile * file,
+	struct pmemfile_block *block)
+{
+	if (block != NULL)
+		return D_RW(block->next);
+	else
+		return file->vinode->first_block;
+}
+
+enum cpy_direction { read_from_blocks, write_to_blocks };
+
+static void
+read_block_range(const struct pmemfile_block *block,
+	uint64_t offset, uint64_t len, char *buf)
+{
+	ASSERT(len > 0);
+	ASSERT(block == NULL || offset < block->size);
+	ASSERT(block == NULL || offset + len <= block->size);
+
+	/* block == NULL means reading from a hole in a sparse file */
+
+	/*
+	 * !is_block_data_initialized(block) means reading from an
+	 * fallocat-ed region in a file, a region that was allocated,
+	 * but never initialized.
+	 */
+
+	if ((block != NULL) && is_block_data_initialized(block)) {
+		const char *read_from = D_RO(block->data) + offset;
+		memcpy(buf, read_from, len);
+	} else {
+		memset(buf, 0, len);
+	}
+}
+
+static void
+write_block_range(PMEMfilepool *pfp, struct pmemfile_block *block,
+	uint64_t offset, uint64_t len, const char *buf)
+{
+	ASSERT(block != NULL);
+	ASSERT(len > 0);
+	ASSERT(offset < block->size);
+	ASSERT(offset + len <= block->size);
+
+	char *data = D_RW(block->data);
+
+	if (!is_block_data_initialized(block)) {
+		char *start_zero = data;
+		size_t count = offset;
+		if (count != 0) {
+			VALGRIND_ADD_TO_TX(start_zero, count);
+			pmemobj_memset_persist(pfp->pop, start_zero, 0, count);
+			VALGRIND_REMOVE_FROM_TX(start_zero, count);
+		}
+
+		start_zero = data + offset + len;
+		count = block->size - (offset + len);
+		if (count != 0) {
+			VALGRIND_ADD_TO_TX(start_zero, count);
+			pmemobj_memset_persist(pfp->pop, start_zero, 0, count);
+			VALGRIND_REMOVE_FROM_TX(start_zero, count);
+		}
+
+		TX_ADD_FIELD_DIRECT(block, flags);
+		block->flags |= BLOCK_INITIALIZED;
+	}
+
+	VALGRIND_ADD_TO_TX(data + offset, len);
+	pmemobj_memcpy_persist(pfp->pop, data + offset, buf, len);
+	VALGRIND_REMOVE_FROM_TX(data + offset, len);
+}
+
+static void
+iterate_on_file_range(PMEMfilepool *pfp, PMEMfile *file,
+    uint64_t offset, uint64_t len, char *buf, enum cpy_direction dir)
+{
+	struct pmemfile_block *block = file_find_block(file, offset);
+
+	while (len > 0) {
+		/* Remember the pointer to block used last time */
+		if (block != NULL)
+			file->block_pointer_cache = block;
+		else
+			ASSERT(dir == read_from_blocks);
```

Comment:
What is this code doing? And why only an assert here? I would think if its critical to understand the cpy_direction we would want to check for that in non-debug builds as well?